src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -167,19 +170,29 @@ internal sealed class VisualStudioProject
                 documentAlreadyInWorkspace: (s, d) => s.ContainsDocument(d),
                 documentAddAction: (w, d) => w.OnDocumentAdded(d),
                 documentRemoveAction: (w, documentId) => w.OnDocumentRemoved(documentId),
-                documentTextLoaderChangedAction: (w, d, loader) => w.OnDocumentTextLoaderChanged(d, loader));
+                documentTextLoaderChangedAction: (s, d, loader) => s.WithDocumentTextLoader(d, loader, PreservationMode.PreserveValue),
+                documentChangedWorkspaceKind: WorkspaceChangeKind.DocumentChanged);
 
             _additionalFiles = new BatchingDocumentCollection(this,
                 (s, d) => s.ContainsAdditionalDocument(d),
                 (w, d) => w.OnAdditionalDocumentAdded(d),
                 (w, documentId) => w.OnAdditionalDocumentRemoved(documentId),
-                documentTextLoaderChangedAction: (w, d, loader) => w.OnAdditionalDocumentTextLoaderChanged(d, loader));
+                documentTextLoaderChangedAction: (s, d, loader) => s.WithAdditionalDocumentTextLoader(d, loader, PreservationMode.PreserveValue),
+                documentChangedWorkspaceKind: WorkspaceChangeKind.AdditionalDocumentChanged);
 
             _analyzerConfigFiles = new BatchingDocumentCollection(this,
                 (s, d) => s.ContainsAnalyzerConfigDocument(d),
                 (w, d) => w.OnAnalyzerConfigDocumentAdded(d),
                 (w, documentId) => w.OnAnalyzerConfigDocumentRemoved(documentId),
-                documentTextLoaderChangedAction: (w, d, loader) => w.OnAnalyzerConfigDocumentTextLoaderChanged(d, loader));
+                documentTextLoaderChangedAction: (s, d, loader) => s.WithAnalyzerConfigDocumentTextLoader(d, loader, PreservationMode.PreserveValue),
+                documentChangedWorkspaceKind: WorkspaceChangeKind.AnalyzerConfigDocumentChanged);
+
+            _fileChangesToProcess = new AsyncBatchingWorkQueue<string>(
+                TimeSpan.FromMilliseconds(200), // 200 chosen with absolutely no evidence whatsoever
+                ProcessFileChangesAsync,
+                StringComparer.Ordinal,
+                workspace.Services.GetRequiredService<IWorkspaceAsynchronousOperationListenerProvider>().GetListener(),
+                _asynchronousFileChangeProcessingCancellationTokenSource.Token);
```

Comment:
normally we listen to the shutdown token.  is this the right token to use here?

Reply:
(replied to https://github.com/dotnet/roslyn/pull/59441#discussion_r803332590)